R/bundlePackage.R

```diff
@@ -25,135 +25,59 @@ bundlePackages <- function(appDir,
-    info <- c(info, as.list(deps[i, grep("Github", colnames(deps), perl = TRUE, value = TRUE)]))
-
-    # get package description; note that we need to remove the
-    # packageDescription S3 class from the object or jsonlite will refuse to
-    # serialize it when building the manifest JSON
-    # TODO: should we get description from packrat/desc folder?
-    info$description <- suppressWarnings(unclass(utils::packageDescription(name)))
-
-    # TODO(HW): should this code be removed? I don't see how it is ever reached
-    # if description is NA, application dependency may not be installed
-    if (is.na(info$description[1])) {
-      errorMessages <- c(
-        errorMessages,
-        paste0(
-          capitalize(assetTypeName), " depends on package \"", name, "\"",
-          "but it is not installed. Please resolve before continuing."
-        )
-      )
-      next
-    }
-
-    # validate package source (returns a message if there is a problem)
-    packageMessages <- c(packageMessages, validatePackageSource(deps[i, ]))
-
-    # good to go
-    packages[[name]] <- info
+  not_installed <- !vapply(deps$Package, is_installed, logical(1))
+  if (any(not_installed)) {
+    pkgs <- deps$Package[not_installed]
+    cli::cli_abort(c(
+      "All packages used by {assetTypeName} must be installed.",
+      x = "Missing packages: {.pkg {pkgs}}."
+    ))
   }
 
-  if (length(packageMessages)) {
-    # Advice to help resolve installed packages that are not available using the
-    # current set of configured repositories. Each package with a missing
-    # repository has already been printed (see snapshotDependencizes).
-    #
-    # This situation used to trigger an error (halting deployment), but was
-    # softened because:
-    #   * CRAN-archived packages are not visible to our available.packages
-    #     scanning.
-    #   * Source-installed packages may be available after a manual server-side
-    #     installation.
-    #
-    # That said, an incorrectly configured "repos" option is almost always the
-    # cause.
-    packageMessages <- c(
-      packageMessages,
-      paste0(
-        "Unable to determine the source location for some packages. ",
-        "Packages should be installed from a package repository like ",
-        "CRAN or a version control system. Check that ",
-        "options('repos') refers to a package repository containing ",
-        "the needed package versions."
+  missing_repo <- !isSCMSource(deps$Source) & is.na(deps$Repository)
+  if (any(missing_repo)) {
+    pkgs <- deps$Package[missing_repo]
+    cli::cli_warn(c(
+      "Packages require a known repository for install on remote system.",
+      x = "Packages with unknown repository: {.pkg {pkgs}}.",
+      i = paste0(
+        "Automatic source detection relies on packages being installed from ",
+        "a standard repository like CRAN or BioConductor, or from a version ",
+        "control system like GitHub or GitLab."
```

Comment:
I don't know what "Automatic source detection" means. Maybe:

> Packages must be installed from a standard repository like CRAN or BioConductor, or from a version control system like GitHub or GitLab.

It's still not great. We're trying to communicate that the _local_ package installs need to come from a supported location so that the same version of that package can be installed by the _remote_ system. These errors often throw folks off because they forget that there are client-side requirements that need to be met before the remote system is ever involved. We occasionally have folks try to install packages on the remote system rather than adjusting their client library.

Should we enumerate the most common mistakes?

* Installing packages from source.
* No repository (`options('repos')`) containing the target package.

Reply:
Yeah, "source" is ambiguous here, e.g. "the source from which a package was installed" vs short-hand for "source code". This is tricky to communicate succinctly.

Maybe something that gives the user more of a clear instruction for how to fix the problem would read better.

> Cannot determine the installation source for these packages. Try reinstalling from a standard…

Reply:
It doesn't seem possible to give enough detail in this error message, so I've gone for a pointer to the docs:

      ! Local packages must be installed from a supported source.
      x Unsupported packages: shiny.
      i Supported sources are CRAN and CRAN-like repositories, BioConductor, GitHub, GitLab, and Bitbucket.
      i See `rsconnect::appDependencies()` for more details.

I'll keep working on the language but this feels like a more fruitful approach.